api/extension-guides/notebook.md

```diff
@@ -57,122 +56,128 @@ A content provider is declared in `package.json` under the `contributes.notebook
+    private _executionOrder = 0;
 
-* [GitHub Issues Notebook](https://github.com/microsoft/vscode-github-issue-notebooks/blob/main/src/extension/notebookProvider.ts): Kernel to execute queries for GitHub Issues
+    constructor() {
+        this._controller = vscode.notebooks.createNotebookController(this.controllerId, this.notebookType, this.label);
 
-<!-- - [HTTP Request Notebook](): Kernel to issue HTTP requests (TODO: PR against https://github.com/Huachao/vscode-restclient to add notebooks) -->
+        this._controller.supportedLanguages = this.supportedLanguages;
+        this._controller.supportsExecutionOrder = true;
+        this._controller.executeHandler = this._execute.bind(this);
+    }
 
-### Best practices
+    private _execute(cells: vscode.NotebookCell[], _notebook: vscode.NotebookDocument, _controller: vscode.NotebookController): void {
+        for (let cell of cells) {
+            this._doExecution(cell);
+        }
+    }
 
-While a kernel need only return an output, it can be desirable to set metadata on cells as it executes them to enable features like the run duration counter, execution order badge, and run status icon. For instance, a kernel's `executeCell` function might look like this:
+    private async _doExecution(cell: vscode.NotebookCell): Promise<void> {
+        const execution = this._controller.createNotebookCellExecution(cell);
+        execution.executionOrder = ++this._executionOrder;
+        execution.start(Date.now()); // Keep track of elapsed time to execute cell.
 
-```ts
-async function executeCell(document: vscode.NotebookDocument, cell: vscode.NotebookCell, token: vscode.CancellationToken) {
-    try {
-        cell.metadata.runState = vscode.NotebookCellRunState.Running;
-        const start = +new Date();
-        cell.metadata.runStartTime = start;
-        cell.metadata.executionOrder = ++this.runIndex;
-        const result = await doExecuteCell(document, cell, token);
-        cell.outputs = [result];
-        cell.metadata.runState = vscode.NotebookCellRunState.Success;
-        cell.metadata.lastRunDuration = +new Date() - start;
-    } catch (e) {
-        cell.outputs = [{ outputKind: vscode.CellOutputKind.Error, ename: e.name, evalue: e.message, traceback: [e.stack] }];
-        cell.metadata.runState = vscode.NotebookCellRunState.Error;
-        cell.metadata.lastRunDuration = undefined;
+        /* Do some execution here; not implemented */
+
+        execution.replaceOutput([new vscode.NotebookCellOutput([vscode.NotebookCellOutputItem.text('Dummy output text!')])])
+        execution.end(true, Date.now());
     }
-};
+}
 ```
 
+Samples:
+
+* [GitHub Issues Notebook](https://github.com/microsoft/vscode-github-issue-notebooks/blob/main/src/extension/notebookProvider.ts): Controller to execute queries for GitHub Issues
+* [REST Book](https://github.com/tanhakabir/rest-book/blob/main/src/extension/notebookKernel.ts): Controller to run REST queries.
+
```

Comment:
another, very simple sample is the regexp notebook: https://github.com/jrieken/vscode-regex-notebook/blob/master/src/extension/extension.ts#L56